runtime/lib/capi/ExecutionContext.hpp

```diff
@@ -59,6 +60,61 @@ class MemoryManager final {
     bool contains(void *ptr) { return _impl.contains(ptr); }
 };
 
+class SharedLibraryManager final {
+  private:
+    void *_handler{NULL};
+
+  public:
+    SharedLibraryManager() = delete;
+    SharedLibraryManager(std::string filename)
+    {
+        // RTLD_DEEPBIND is incompatible with sanitizers.
+        // If you have compiled this file with sanitizers and you reach this line
+        // you will get an error.
+        // Please re-compile without sanitizers.
+        _handler = dlopen(filename.c_str(), RTLD_LAZY | RTLD_DEEPBIND);
+        RT_FAIL_IF(!_handler, dlerror());
+    }
+
+    ~SharedLibraryManager()
+    {
+        // dlopen and dlclose increment and decrement reference counters.
+        // Since we have a guaranteed _handler in a valid SharedLibraryManager instance
+        // then we don't really need to worry about dlclose.
+        // In other words, there is a one to one correspondance between an instance
+        // of SharedLibraryManager and an increase in the reference count for the dynamic library.
+        // dlclose returns non-zero on error.
+        //
+        // Errors in dlclose are implementation dependent.
+        // There are two possible errors during dlclose in glibc: "shared object not open"
+        // and "cannot create scope list". Look for _dl_signal_error in:
+        //
+        //     codebrowser.dev/glibc/glibc/elf/dl-close.c.html/
+        //
+        // This means that at the very least, one could trigger an error in the following line by
+        // doing the following: dlopen the same library and closing it multiple times in a different
+        // location.
+        //
+        // This would mean that the reference count would be less than the number of instances
+        // of SharedLibraryManager.
+        //
+        // There really is no way to protect against this error, except to always use
+        // SharedLibraryManager to manage shared libraries.
+        //
+        // Exercise for the reader, how could one trigger the "cannot create scope list" error?
+        dlclose(_handler);
+    }
+
+    void *getSymbol(std::string symbol)
```

Comment:
```suggestion
    void *getSymbol(const std::string &symbol)
```

Reply:
Changed without adding suggestion to batch.